### PR TITLE
[feature] support convert alpha rowset

### PR DESCRIPTION
**Pull request by jacktengg**
# Proposed changes

This is a part of https://github.com/apache/incubator-doris/issues/8984

Code of AlphaRowset  is going to be abolished :
- Apache Doris has set the default rowset to BetaRowset in version 0.15 for about half a year.
- In version 1.0.0, we released a vectorized execution engine, in which the storage vectorization is only designed and modified for BetaRowset, so the old AlphaRowset will not enjoy the huge performance improvement brought by vectorization, and we have no plans to do vectorization-related optimizations for AlphaRowset in the future. 
- In version 1.1, users will be prohibited from creating AlphaRowset tables. Existing tables can be automatically converted to beta rowset if there are alpha rowsets.
- We will completely remove the AlphaRowset related code in version 1.2, so that the storage side of Doris can iterate quickly. For the existing AlphaRowset table, we provide the following methods for data conversion in 1.1:

1. Check whether the alpha rowset exists on all current BEs, use the following SQL: `admin show tablet storage format`
2. If there is an alpha rowset on the BE, add the `convert_rowset_thread_num` configuration item in the BE configuration file. After restarting the BE, the BE will automatically convert the alpha rowset to beta rowset.

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/convert_rowset.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#include "olap/merger.h"
+#include "olap/storage_engine.h"
+#include "olap/tablet.h"
+
+namespace doris {
+class DataDir;
+class ConvertRowset {
+public:
+    ConvertRowset(TabletSharedPtr tablet) : _tablet(tablet) {}
+    Status prepare_convert();
```
Comment (yiguolei) [inline be/src/olap/convert_rowset.h:29]: 这里面没这么多函数，我们不要prepare，convert 这一堆了。 因为我们就一个rowset，所以也不会有很多reader

File: be/src/olap/convert_rowset.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                      << ", current_max_version=" << current_max_version
+                      << ", disk=" << _tablet->data_dir()->path() << ", segments=" << _input_rowsets[i]->num_segments()
+                      << ". elapsed time=" << watch.get_elapse_second()
+                      << "s.";
+        }
+    }
+    return Status::OK();
+}
+
+Status ConvertRowset::construct_input_rowset_readers() {
```
Comment (yiguolei) [inline be/src/olap/convert_rowset.cpp:107]: 这个函数不要了，因为我们就一个rowset，那么就在调用的地方直接获取这一个rowset 的reader 就好了

File: be/src/olap/convert_rowset.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    Status do_convert();
+
+private:
+    Status construct_output_rowset_writer();
```
Comment (yiguolei) [inline be/src/olap/convert_rowset.h:33]: construct_input 和output rowset 都没用了，里面都就1行代码，不要单独的函数了

File: be/src/olap/convert_rowset.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    Status construct_input_rowset_readers();
+    void pick_rowsets_to_convert();
+    Status check_correctness(RowsetSharedPtr input_rowset, RowsetSharedPtr output_rowset,
+                             const Merger::Statistics& stats);
+    int64_t _get_input_num_rows_from_seg_grps(RowsetSharedPtr rowset);
+    void _modify_rowsets(RowsetSharedPtr input_rowset, RowsetSharedPtr output_rowset);
+
+private:
+    TabletSharedPtr _tablet;
+    std::vector<RowsetSharedPtr> _input_rowsets;
```
Comment (yiguolei) [inline be/src/olap/convert_rowset.h:43]: input rowsets和input rs readers 也不要了，

File: be/src/olap/convert_rowset.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        return Status::OLAPInternalError(OLAP_ERR_CE_TRY_CE_LOCK_ERROR);
+    }
+
+    std::vector<RowsetSharedPtr> alpah_rowsets;
+    _tablet->find_alpha_rowsets(&alpah_rowsets);
+
+    Merger::Statistics stats;
+    Status res;
+    for (size_t i = 0; i < alpah_rowsets.size(); ++i) {
+        Version output_version =
```
Comment (yiguolei) [inline be/src/olap/convert_rowset.cpp:39]: 这里不能一直跑， 这里可以记录一下我们写过多少个rowset，或者merge过多少行记录，就得退出，否则正常的compaction没法跑了

File: be/src/olap/olap_server.cpp [hunk 6]
```diff
@@ -304,6 +320,27 @@ void StorageEngine::_tablet_checkpoint_callback(const std::vector<DataDir*>& dat
     } while (!_stop_background_threads_latch.wait_for(std::chrono::seconds(interval)));
 }
 
+void StorageEngine::_alpha_rowset_scan_thread_callback() {
+    LOG(INFO) << "try to start alpha rowset scan thread!";
+
+    do {
+        std::vector<TabletSharedPtr> tablet_have_alpha_rowset;
+        _tablet_manager->find_tablet_have_alpha_rowset(tablet_have_alpha_rowset);
```
Comment (yiguolei) [inline be/src/olap/olap_server.cpp:328]: 这中间shuffle 一下

File: be/src/olap/olap_server.cpp [hunk 7]
```diff
@@ -304,6 +320,27 @@ void StorageEngine::_tablet_checkpoint_callback(const std::vector<DataDir*>& dat
...
+        for (int i = 0; i < tablet_have_alpha_rowset.size(); ++i) {
```
Comment (yiguolei) [inline be/src/olap/olap_server.cpp:329]: 这里不要一下submit 这么多task，我们最多submit 线程数 * 2 个task；现在这个写法会重复submit 很多相同的tablet 到队列里啊。

Comment (yiguolei) [reply be/src/olap/olap_server.cpp:329]: 我觉得submit 完最多线程数个task之后，就wait 他们都结束，然后再submit 下一堆

File: be/src/olap/olap_server.cpp [hunk 8]
```diff
@@ -72,6 +73,21 @@ Status StorageEngine::start_bg_threads() {
             .set_max_threads(max_thread_num)
             .build(&_compaction_thread_pool);
 
+    int32_t convert_rowset_thread_num = config::convert_rowset_thread_num;
+    if (convert_rowset_thread_num > 0) {
+        // alpha rowset scan thread
+        RETURN_IF_ERROR(Thread::create(
+                "StorageEngine", "alpha_rowset_scan_thread",
+                [this]() { this->_alpha_rowset_scan_thread_callback(); },
```
Comment (yiguolei) [inline be/src/olap/olap_server.cpp:81]: 启动scan 线程和 初始化thread pool 是不是需要调换一下顺序，否则scan thread 启动了，但是pool没初始化可能会挂啊

File: be/src/olap/olap_server.cpp [hunk 9]
```diff
@@ -304,6 +320,49 @@ void StorageEngine::_tablet_checkpoint_callback(const std::vector<DataDir*>& dat
...
+    std::vector<DataDir*> data_dirs;
+    for (auto& tmp_store : _store_map) {
+        data_dirs.push_back(tmp_store.second);
+    }
+
+    auto scan_interval_sec = config::scan_alpha_rowset_min_interval_sec;
+    auto max_convert_task = config::convert_rowset_thread_num * 2;
+    do {
+        std::vector<TabletSharedPtr> tablet_have_alpha_rowset;
+        _tablet_manager->find_tablet_have_alpha_rowset(data_dirs, tablet_have_alpha_rowset);
```
Comment (yiguolei) [inline be/src/olap/olap_server.cpp:335]: 这里不要传入data dirs参数了

**Review by yiguolei — APPROVED** [review]
LGTM

**Comment by morningman** [comment]
Hi @jacktengg , Could you please push this PR to dev-1.0.1?

**Pull request merged**